core/src/util/usFrameworkFactory.cpp

```diff
@@ -27,10 +27,46 @@
 
 namespace us {
 
+// A helper class to shut down the framework when all external
+// references (Bundle, BundleContext, etc.) are destroyed. It
+// also manages the lifetime of the CoreBundleContext instance.
+struct CoreBundleContextHolder
+{
+  CoreBundleContextHolder(std::unique_ptr<CoreBundleContext> ctx)
+    : ctx(std::move(ctx))
+  {}
+
+  ~CoreBundleContextHolder()
+  {
+    auto const state = ctx->systemBundle->state.load();
+    if (((Bundle::STATE_STARTING | Bundle::STATE_ACTIVE) & state) == 0)
+    {
+      // Call WaitForStop in case some did call Framework::Stop()
+      // but didn't wait for it. This joins with a potentially
+      // running framework shut down thread.
+      ctx->systemBundle->WaitForStop(std::chrono::milliseconds::zero());
+      return;
+    }
+
+    // Create a new CoreBundleContext holder, in case some event listener
+    // gets the system bundle and starts it again.
+    auto fwCtx = ctx.get();
+    std::shared_ptr<CoreBundleContext> holder(std::make_shared<CoreBundleContextHolder>(std::move(ctx)), fwCtx);
+    holder->SetThis(holder);
+    holder->systemBundle->Shutdown(false);
```

Comment:
**_The following is not caused by these changes, however I've noticed a different issue in the same area as these changes which I'd like to discuss._**

I've run into a crash on Windows when a ```Framework``` instance is stored in a DLL as a global. On static destruction (assuming ```Framework::Stop()``` hasn't already been called), the ```std::thread``` creation inside ```FrameworkPrivate::Shutdown``` will throw an exception. This happens because while in DllMain, you shouldn't do much of anything (as stressed in [Windows documentation](https://msdn.microsoft.com/en-us/library/windows/desktop/dn633971(v=vs.85).aspx#general_best_practices)), including spawning threads. Of course this only occurs if no one calls ```Framework::Stop``` prior to static destruction.

While I don't want to condone the over-use of global variables/singletons, I still think its worth it to try and make sure static destruction occurs smoothly. Within our use at the MathWorks we only want a single Framework instance for the entire process (for now).

What are your thoughts on making a "synchronous shutdown" which is only called during implicit Framework destruction (i.e. when a Framework instance goes out of scope)? Explicit calls to ```Framework::Stop()``` would still be asynchronous.